rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
+
+Side effect: we would get our envisioned convert operator almost for free:
+
+```bash
+vast exec 'parse zeek | print csv'
+# Obviates the need for:
+vast exec 'convert zeek csv'
+```
+
+Another UX improvement would be assuming that we interact with a VAST node. Here
+we could have dedicate commands the prepend or append a pipeline operator:
+
+- `vast pull` → `from vast://...`
+- `vast push` → `to vast://...`
+
+An `export` would then become:
+
+```bash
+vast pull 'where EXPRESSION | print json'
+# Same with exec:
+vast exec 'from vast://.... | where EXPRESSION | print json'
+# Same with export:
+vast export json EXPRESSION
+```
+
+An `import` degenerates to
+
+```bash
+vast push 'parse zeek | where EXPRESSION'
+# Same with exec
+vast exec 'parse zeek | where EXPRESSION | to vast://...'
+# Same with import
+vast import zeek EXPRESSION
+```
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the middle.
+It's clean and simple.
+
+But this may be too restrictive. Let's assume for a moment that input and output
+are dynamically typed. Then an operator must define its input and output types,
+e.g., `from<In, Out>`. This can come in handy for working with matchers:
+
+Here, we build a matcher, with read having the type `read<Bytes, Arrow>`:
+
+```bash
+vast exec 'read csv | matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+The type of `matcher build` would be `<Arrow, Flatbuffer>`. Also assume `matcher
+load` yields `<Flatbuffer, Void>` where `Void` is "no output, just a side
+effect".
+
+How do we pre-load a matcher?
+
+```bash
+vast exec 'from file:///ips.flatbuffer |
+           matcher load --name=ips vast:///1.2.3.4'
+```
+
+Or more UNIX'ish, assuming an implicit `from -`:
+
+```bash
+vast exec 'matcher load --name=ips' < ips.flatbuffer
+```
+
+The `matcher` sub-command `match` does the actual matching:
+
+```bash
+vast pull 'where #type == "suricata.flow" |
+           matcher match --on=:addr ips'
+```
+
+The above is a locally running pipeline, but most likely I want this to run at a
+node. This begs the challenge: how do we manage all these pipelines? With a
+client command (and ideally also a REST API):
+
+```bash
+# Show all pipelines:
+vast pipeline list
+```
+
+Create remote pipelines at ease:
+
+```bash
+vast pipeline create foo 'where #type == "suricata.flow" | matcher match ips'
+vast pipeline list
+# Output:
+# foo: where #type == "suricata.flow" | matcher match ips
+```
+
+The idea would be that only clients can manipulate server state. This fixes
+the mess of having server-side YAML that goes out of sync with the server
+state when new things happen, e.g., new schemas arrive or new matchers get
+spawned. From now on, everything is client-side managed. This is where you can
+have your YAML for declarative ops:
+
+```bash
+vast pipeline load < pipelines.yaml
```

Comment:
Would VAST auto-reload pipelines.yaml when it changes? Or would there be a command to do that without stopping the server? What do you expect the default behavior to be between being to apply a new pipelines.yaml file only when the server is done, allowing reload with command and VAST monitoring the file itself to autoreload the config?

Reply:
I added a sentence to the state management section. Let me know if this needs further clarification.

Basically the idea is that only client commands can change the server-side state. Dumping the server-side state is equivalent to executing a set of client commands to add, remove, update pipelines.

Reply:
NB: I think for the RFC review workflow it'd be much nicer if you proposed your change as a comment, and then only pushed it once we resolved the thread.

Reply:
This may work for smaller changes only, but yeah, we could try.

Reply:
Thanks, I think that makes sense :-) (feel free to resolve the thread)